core/src/main/java/org/javarosa/core/model/instance/TreeElement.java

```diff
@@ -671,233 +720,106 @@ public void setAttribute(String namespace, String name, String value) {
+            attributes = null;
+        } else {
+            attributes = new Vector<TreeElement>();
+            int attrCount = (int)ExtUtil.readNumeric(in);
+            for (int i = 0; i < attrCount; ++i) {
+                TreeElement attr = new TreeElement();
+                attr.readExternal(in, pf);
+                attr.setParent(this);
+                attributes.addElement(attr);
+            }
+        }
     }
 
-    /*
-     * (non-Javadoc)
-     *
-     * @see
-     * org.javarosa.core.services.storage.utilities.Externalizable#writeExternal
-     * (java.io.DataOutputStream)
-     */
-    /* (non-Javadoc)
-     * @see org.javarosa.core.model.instance.AbstractTreeElement#writeExternal(java.io.DataOutputStream)
-     */
+    @Override
     public void writeExternal(DataOutputStream out) throws IOException {
         ExtUtil.writeString(out, ExtUtil.emptyIfNull(name));
         ExtUtil.writeNumeric(out, multiplicity);
         ExtUtil.writeNumeric(out, flags);
         ExtUtil.write(out, new ExtWrapNullable(value == null ? null : new ExtWrapTagged(value)));
 
-        // Jan 22, 2009 - csims@dimagi.com
-        // old line: ExtUtil.write(out, new
-        // ExtWrapList(ExtUtil.emptyIfNull(children)));
-        // New Child serialization
-        // 1. write null status as boolean
-        // 2. write number of children
-        // 3. for all child in children
-        // 3.1 if child type == TreeElement write boolean true , then serialize
-        // directly.
-        // 3.2 if child type != TreeElement, write boolean false, then tagged
-        // child
+        writeChildrenToExternal(out);
+
+        ExtUtil.writeNumeric(out, dataType);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(instanceName));
+        ExtUtil.write(out, new ExtWrapNullable(constraint)); // TODO: inefficient for repeats
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(preloadHandler));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(preloadParams));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(namespace));
+
+        writeAttributesToExternal(out);
+    }
+
+    private void writeChildrenToExternal(DataOutputStream out) throws IOException {
         if (children == null) {
-            // 1.
             ExtUtil.writeBool(out, false);
         } else {
-            // 1.
             ExtUtil.writeBool(out, true);
-            // 2.
             ExtUtil.writeNumeric(out, children.size());
-            // 3.
             Enumeration en = children.elements();
             while (en.hasMoreElements()) {
                 TreeElement child = (TreeElement)en.nextElement();
-                if (child.getClass() == TreeElement.class) {
-                    // 3.1
-                    ExtUtil.writeBool(out, true);
-                    child.writeExternal(out);
-                } else {
-                    // 3.2
-                    ExtUtil.writeBool(out, false);
-                    ExtUtil.write(out, new ExtWrapTagged(child));
```

Comment:
Note, I took out the `if (child.getClass() == TreeElement.class) {` code while I was at it, since we are doing a migration anyways.